service/tablet_allocator.cc

```diff
@@ -3216,37 +3265,112 @@ class load_balancer {
+                    shard_load_info.tablet_sizes_per_table[table] += tablet_sizes_sum;
                     node_load_info.tablet_count_per_table[table] += tids.size();
-                    total_load += tids.size();
+                    node_load_info.tablet_sizes_per_table[table] += tablet_sizes_sum;
+                    if (node_load_info.dusage) {
+                        node_load_info.dusage->used += tablet_sizes_sum;
+                    }
+                    total_tablet_count += tids.size();
+                    total_tablet_sizes += tablet_sizes_sum;
                     if (tmap.needs_merge() && tids.size() == 2) {
                         // Exclude both sibling tablets if either haven't finished migration yet. That's to prevent balancer from
                         // un-doing the colocation.
                         if (!migrating(t1) && !migrating(t2)) {
                             auto candidate = colocated_tablets{global_tablet_id{table, t1.tid}, global_tablet_id{table, t2->tid}};
-                            add_candidate(shard_load_info, migration_tablet_set{std::move(candidate)});
+                            add_candidate(shard_load_info, migration_tablet_set{std::move(candidate), tablet_sizes_sum});
                         }
                     } else {
-                        for (auto tid : tids) {
-                            if (!migrating(get_table_desc(tid))) { // migrating tablets are not candidates
-                                add_candidate(shard_load_info, migration_tablet_set{global_tablet_id{table, tid}});
+                        if (tids.size() != tablet_sizes.size()) {
+                            on_internal_error(lblogger, "Number of co-located tablets and their sizes don't match.");
+                        }
+                        for (size_t i = 0; i < tids.size(); i++) {
+                            if (!migrating(get_table_desc(tids[i]))) { // migrating tablets are not candidates
+                                add_candidate(shard_load_info, migration_tablet_set{global_tablet_id{table, tids[i]}, tablet_sizes[i]});
                             }
                         }
                     }
                 }
 
                 return make_ready_future<>();
             });
-            _tablet_count_per_table[table] = total_load;
+            _disk_used_per_table[table] = total_tablet_sizes;
+            _tablet_count_per_table[table] = total_tablet_count;
+        }
+
+        // Compute load imbalance.
+
+        _total_capacity_shards = 0;
+        _total_capacity_nodes = 0;
+        _total_capacity_storage = 0;
+        load_type max_load = 0;
+        load_type min_load = 0;
+        std::optional<host_id> min_load_node = std::nullopt;
+        for (auto&& [host, load] : nodes) {
+            load.update();
+            _stats.for_node(dc, host).load = load.avg_load;
+
+            if (!load.drained && (_load_sketch->has_complete_data(host) || !nodes_to_drain.empty())) {
```

Comment:
This logic for node exclusion is duplicated in several places. We should do it by dropping nodes from "nodes", like with `skip_nodes`. Earlier.

Reply:
> skip_nodes

I meant `_skiplist`.

The code is here:

```
        // Apply skiplist only when not draining.
        // It's unsafe to move tablets to non-skip nodes as this can lead to node overload.
        if (nodes_to_drain.empty()) {
            for (auto host_to_skip : _skiplist) {
                if (auto handle = nodes.extract(host_to_skip)) {
                    auto& node = handle.mapped();
                    lblogger.debug("Ignoring dead node {}: state={}", node.id, node.node->get_state());
                }
            }
        }
```

What this accomplishes is that `nodes` contain only the set of nodes to balance. On which algorithms rely later,  they check if a node is contained in `nodes` to determine if it's in the balanced set.